vlm/src/main/scala/geotrellis/contrib/vlm/RasterSource.scala

```diff
@@ -94,7 +98,8 @@ trait RasterSource extends CellGrid[Long] with Serializable {
     * @group reproject
     */
   def reproject(crs: CRS, method: ResampleMethod = NearestNeighbor, strategy: OverviewStrategy = AutoHigherResolution): RasterSource =
-    reproject(crs, Reproject.Options(method = method), strategy)
+    if (crs == this.crs) this
+    else reproject(crs, Reproject.Options(method = method), strategy)
```

Comment:
The PR does not pass the CI yet, but when it does I'm pretty sure this is going to be a circular reference that blows the stack.

Thinking about this I think there are two things that make me feel weird about the protected approach

1. `_reproject` - the underscore convention universally means "private variable, don't pay attention to this"  so it's weird to be asking people to implement it.
2. `_reproject` - name looks pretty similar to `reproject` that can easily lead to careless mistake like the one here thus adding a little bit of misery to API user.

The only name I can think of is `reprojectInternal` which ... is different but not sure is super meaningful.
++ @pomadchin for naming discussion
++ @metasim as our start external user for this API

Reply:
I agree that `_reproject` is a bad name. Probably it is because of me it appeared in this PR ([since I mentioned it for some reason here](https://github.com/geotrellis/geotrellis-contrib/issues/102#issuecomment-509371493) just to give a better task description).

Is there an `underscore convention` in scala? As I it is said [here](https://docs.scala-lang.org/style/naming-conventions.html) it looks like it is a convention for `mutators`.

Names that come to my mind are `reprojectBase`, `reprojectRaw`, `reprojectUnsafe`, `reprojectForce`

Reply:
I agree that `_reproject` isn't really the best name, but I was unable to come up with anything better. I also don't think the `under score convention` is supposed to be used this way in Scala. However, I noticed that we use this style in some places in GeoTrellis

https://github.com/locationtech/geotrellis/blob/master/raster/src/main/scala/geotrellis/raster/io/geotiff/GeoTiffMultibandTile.scala#L811

https://github.com/locationtech/geotrellis/blob/master/spark/src/main/scala/geotrellis/spark/store/LayerWriter.scala#L124

Reply:
In case of a `GeoTiffMultibandTile.scala` it is okay (IMO) since nobody would ever use it / overload it in the client API.

In case of LayerWriter it is an oversight.

Reply:
This PR looks good but I'm not really comfortable with something being called `internal` which is actually required of any would-be implementers. We can either find a better name (struggling to do that, personally) or expect implementers to provide their own short circuiting logic within `reproject`